### PR TITLE
Configure /proc with hidepid by default and add doPseudoFsMount to addEntryToFstab

### PR DESCRIPTION
**Pull request by rlmenge**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

`Hidepid` is a mount option that prevents proc ids from being visible to all users. Without it, a non-privileged user can see all process identifiers (PIDs) and can `cat` the pid to see which process is involved.

Therefore, set ` hidepid` by default and allow users to opt out using the `HidepidDisable` config option. In order for these options to persist, the mount point is added to the fstab file. 

Since `/proc` is a pseudo file system, an argument was added to `addEntryToFstab` that specifies whether the mountpoint relates to a pseudo file system (such as /proc, /sys). This argument allows `device` to be set to the given path and the `pass num` be set to `0` since fsck checking does not apply to pseudo filesystems.  

Test with following commands
To see pids (with hidepid will only show those associated with your user unless privileged)
`ls -ld /proc/[0-9]*`

To see updated fstab file
`cat /etc/fstab`

To see /proc options
`mount | grep ^proc`

References:
https://linux-audit.com/linux-system-hardening-adding-hidepid-to-proc/
https://www.iezzi.ch/process-hiding-hidepid-capabilities-of-procfs/ 

###### Change Log  

- Added bool json config option "HidepidDisable". When set to true, proc is not mounted with the hidepid option and is not added to the fstab file. This results in hidepid being set by default
- Updated updateFstab to check if hidepid is enabled. If so, calls addEntryToFstab 
- Updated addEntryToFstab to check if using pseudo filesystem mount and if so, to disable pass and use given device
- Updated documentation

###### Does this affect the toolchain?  

NO

###### Test Methodology

- local build (x86, vhdx in hyper-v)

### CODE REVIEW

File: toolkit/tools/imager/imager.go [hunk 1]
```diff
@@ -481,6 +481,12 @@ func buildImage(mountPointMap, mountPointToFsTypeMap, mountPointToMountArgsMap m
 		setupChrootPackages = append(setupChrootPackages, toolingPackage.Name)
 	}
 
+	logger.Log.Infof("HidepidDisabled is %v.", systemConfig.HidepidDisable)
+	hidepidEnable := true
```
Comment (PawelWMS) [inline toolkit/tools/imager/imager.go:485]: Nit: to make it shorter you can also do `hidepidEnable := !systemConfig.HidepidDisable`.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 2]
```diff
@@ -761,21 +762,28 @@ func updateInitramfsForEncrypt(installChroot *safechroot.Chroot) (err error) {
...
+		err = addEntryToFstab(installRoot, "/proc", "proc", "proc", "rw,nosuid,nodev,noexec,relatime,hidepid=2", true)
+		if err != nil {
+			return
+		}
+	}
 	return
 }
 
-func addEntryToFstab(installRoot, mountPoint, devicePath, fsType, mountArgs string) (err error) {
+func addEntryToFstab(installRoot, mountPoint, devicePath, fsType, mountArgs string, doPseudoFsMount bool) (err error) {
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/installutils/installutils.go:786]: Please also mention the change related to `doPseudoFsMount` in the PR's summary or (even cleaner) move it to a separate PR.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 3]
```diff
@@ -761,21 +762,28 @@ func updateInitramfsForEncrypt(installChroot *safechroot.Chroot) (err error) {
...
 }
 
-func updateFstab(installRoot string, installMap, mountPointToFsTypeMap, mountPointToMountArgsMap map[string]string) (err error) {
+func updateFstab(installRoot string, installMap, mountPointToFsTypeMap, mountPointToMountArgsMap map[string]string, hidepidEnable bool) (err error) {
 	ReportAction("Configuring fstab")
 
 	for mountPoint, devicePath := range installMap {
 		if mountPoint != "" && devicePath != NullDevice {
-			err = addEntryToFstab(installRoot, mountPoint, devicePath, mountPointToFsTypeMap[mountPoint], mountPointToMountArgsMap[mountPoint])
+			err = addEntryToFstab(installRoot, mountPoint, devicePath, mountPointToFsTypeMap[mountPoint], mountPointToMountArgsMap[mountPoint], false)
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/installutils/installutils.go:770]: Instead of passing `false`, please create a const with a name explaining its meaning in the context of that function. I'd avoid directly passing non-obvious arguments like "[magic numbers](https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad)" or boolean values. Look for `const squashErrors = false` for a few examples.

File: toolkit/tools/imagegen/configuration/systemconfig.go [hunk 4]
```diff
@@ -30,6 +30,7 @@ type SystemConfig struct {
 	Encryption         RootEncryption      `json:"Encryption"`
 	RemoveRpmDb        bool                `json:"RemoveRpmDb"`
 	ReadOnlyVerityRoot ReadOnlyVerityRoot  `json:"ReadOnlyVerityRoot"`
+	HidepidDisable     bool                `json:"HidepidDisable"`
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/configuration/systemconfig.go:33]: Please also extend the config tests now to include the new field.

File: toolkit/docs/formats/imageconfig.md [hunk 5]
```diff
@@ -251,6 +251,10 @@ A sample KernelCommandLine enabling a basic IMA mode and passing two additional
 },
 ```
 
+### HidepidDisable
+
+An optional flag that removes the hidepid option from /proc. Hidepid prevents proc ids from being visible to all users.
```
Comment (PawelWMS) [inline toolkit/docs/formats/imageconfig.md:256]: Since this is a user-facing piece of documentation, I'd suggest a few changes:
- Surround "hidepid" in quotes, or use _italic text_, or put it in a `code block`.
- Let's use "IDs" - [with a capital "ID" and a small "s"](https://english.stackexchange.com/questions/212040/is-the-plural-form-of-id-spelled-ids-or-id).
- Maybe also put `/path` in a code block.

Not sure about the last one. Where is our documentation writer?

**Review by PawelWMS — CHANGES_REQUESTED** [review]

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 6]
```diff
@@ -761,21 +762,28 @@ func updateInitramfsForEncrypt(installChroot *safechroot.Chroot) (err error) {
...
 			if err != nil {
 				return
 			}
 		}
 	}
+
+	if hidepidEnable {
+		err = addEntryToFstab(installRoot, "/proc", "proc", "proc", "rw,nosuid,nodev,noexec,relatime,hidepid=2", true)
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/installutils/installutils.go:778]: Won't this be a problem if there already was an entry for `/proc` inside `installMap`?

**Review by PawelWMS — CHANGES_REQUESTED** [review]

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 6 above)
Comment (rlmenge) [reply toolkit/tools/imagegen/installutils/installutils.go:778]: /proc should never be added to the installMap. InstallMap gets its data the following way
```partIDToDevPathMap --> mountPointDevPathMap  == mountPointMap --> installMap``` 
Therefore, since `/proc ` never has a partition, it is not added. That being said, a user could add /proc to the fstab using postinstall scripts which would be a problem. I should document that case.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 7]
```diff
@@ -807,6 +815,8 @@ func addEntryToFstab(installRoot, mountPoint, devicePath, fsType, mountArgs stri
 		device = devicePath
 	} else if diskutils.IsReadOnlyDevice(devicePath) {
 		device = devicePath
+	} else if doPseudoFsMount {
```
[the remark this answers is not in the document]
Comment (rlmenge) [reply toolkit/tools/imagegen/installutils/installutils.go]: Collapsed :)

File: toolkit/tools/imager/imager.go [hunk 8]
```diff
@@ -481,6 +481,12 @@ func buildImage(mountPointMap, mountPointToFsTypeMap, mountPointToMountArgsMap m
 		setupChrootPackages = append(setupChrootPackages, toolingPackage.Name)
 	}
 
+	logger.Log.Infof("HidepidDisabled is %v.", systemConfig.HidepidDisable)
```
[the remark this answers is not in the document]
Comment (rlmenge) [reply toolkit/tools/imager/imager.go]: made it all past tense :)

File: toolkit/tools/imager/imager.go (hunk 1 above)
Comment (rlmenge) [reply toolkit/tools/imager/imager.go]: Done

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 2 above)
Comment (rlmenge) [reply toolkit/tools/imagegen/installutils/installutils.go:789]: Keeping in this PR but did update the title

File: toolkit/tools/imagegen/configuration/systemconfig.go (hunk 4 above)
Comment (rlmenge) [reply toolkit/tools/imagegen/configuration/systemconfig.go]: Added to the large test structures

File: toolkit/docs/formats/imageconfig.md (hunk 5 above)
Comment (rlmenge) [reply toolkit/docs/formats/imageconfig.md]: updated!

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 6 above)
Comment (PawelWMS) [reply toolkit/tools/imagegen/installutils/installutils.go]: Please do, thank you!

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 2 above)
Comment (PawelWMS) [reply toolkit/tools/imagegen/installutils/installutils.go:789]: Could you also extend the summary to explain the need for this change. Is it something the users should know about? Or to ask in a different way: should we mention the effect of this change in the documentation?

Comment (rlmenge) [reply toolkit/tools/imagegen/installutils/installutils.go:789]: At the moment, users should not see this change. I have plans to refactor this function to allow for users to configure this option in the future but that will be in a future PR :).  I can add some more detail to the summary

**Review by PawelWMS — APPROVED** [review]
